Add Markdown documentation (#58)
conda/conda-libmamba-solver: docs/dev-implementation.md

both within the `conda_libmamba_solver` package itself, and as a `conda` plugin.

## Repository structure

* `.github/workflows/`: CI pipelines to run unit and upstream tests, as well as linting and performance benchmarks.
  Some extra workflows might be added by the `conda/infra` settings.
* `conda_libmamba_solver/`: The Python package. Check sections below for details.
* `recipe/`: The conda-build recipe used for the PR build previews. It should be kept in sync with `conda-forge` and `defaults`.
* `dev/`: Supporting scripts and configuration files to set up development environments.
* `docs/`: Documentation sources.
* `tests/`: PyTest testing infrastructure.
* `pyproject.toml`: Project metadata. See below for details.

## Project metadata

The `pyproject.toml` file stores the required packaging metadata, 
as well as some the configuration for some tools (`black`, `pytest`, etc.).

Some peculiarities:

* `flit_core` is the chosen backend for the packaging (as opposed to `setuptools`).
* The `version` is dynamically taken from `conda_libmamba_solver/__init__.py`.
* `black` uses a line length of 99 characters.

> * PyTest configurations are extensive but mainly to deal with upstream testing. 

Reviewer: ```suggestion
* PyTest configurations are extensive but mainly only deal with upstream testing. 
```
...maybe this wording is not totally correct, just trying to clarify what this bullet point is saying

Author: I was trying to say that the pytest configs in pyproject.toml are quite detailed, but they are only needed when we are running the upstream tests. They don't have much of an effect for `conda-libmamba-solver` itself.

Reviewer: Gotcha! Ok so maybe this bullet point can be re-worded like: 
```suggestion
* PyTest configurations are extensive but are only necessary when dealing with upstream testing. 
```
What do you think?

  Check "Development workflows" for details.

## `conda_libmamba_solver` package

The package is a flat namespace:

* `conda_libmamba_solver.__init__`: Defines `__version__` and the old-style plugin API.
* `conda_libmamba_solver.exceptions`: Subclasses of `conda.exceptions`.
* `conda_libmamba_solver.index`: Helper objects to deal with repodata fetching and loading, interfacing with `libmamba` helpers.
* `conda_libmamba_solver.mamba_utils`: Utility functions to help set up the `libmamba` objects.
* `conda_libmamba_solver.models`: Application-agnostic objects to assist in the metadata collection phases.
* `conda_libmamba_solver.solver`: The `conda.core.solve.Solver` subclass with all the libmamba-specific logic.
* `conda_libmamba_solver.state`: Solver-agnostic objects to assist in the solver state specification and collection.
* `conda_libmamba_solver.utils`: Other application-agnostic utility functions.

> Refer to each module docstrings for further details!

### Solver-agnostic parts

The idea behind the module separation is to have better logic reusability and separation between the `libmamba` library and the preparation logic `conda` uses.

> The following paragraphs assume you have read the Deep Dive guides in the `conda` documentation, but as a refresher:

Reviewer: Can we put in a link for the Deep Dive guides here?
